### ISSUE TITLE
False positive property initialization detection

### ISSUE THREAD

**Issue by jtojnar** [vimeo/psalm#2906]
If I initialize a property in a `setup` method, Psalm correctly narrows the type of the property to disallow `null`:

```php
<?php

class Foo {
	/** @var ?int */
	private $val = null;

	private function setup(): void {
		$this->val = 4;
	}

	/**
	 * @return int
	 */
	public function get() {
		if ($this->val === null) {
			$this->setup();
		}

		return $this->val;
	}
}
```

https://psalm.dev/r/5d1e8dacd6

But if I move the `null` check inside the initialization method to reduce the code clutter, it fails to narrow the property’s type:

```php
<?php

class Foo {
	/** @var ?int */
	private $val = null;

	private function setup(): void {
		if ($this->val === null) {
			$this->val = 4;
		}
	}

	/**
	 * @return int
	 */
	public function get() {
		$this->setup();

		return $this->val;
	}
}
```

https://psalm.dev/r/f80aa2b24e

**Comment by weirdan** [comment]
You can express postconditions with `@psalm-assert` on any method, and Psalm will respect those: https://psalm.dev/r/81350aba0f

**Comment by muglug** [comment]
Yup, what he said

**Comment by jtojnar** [comment]
Why does it behave differently if the null check is inside the `setup` function though? It can already infer that the function is setting the property to `!null`.
